lib/util/extractSourceMap.js

```diff
@@ -0,0 +1,252 @@
+ * @param {function (Error|null, {sourceURL: string, sourceContent: string}=): void} callback callback
+ * @returns {void}
+ */
+function fetchFromURL(fs, context, url, sourceRoot, callback) {
+	// 1. It's an absolute url and it is not `windows` path like `C:\dir\file`
+	if (/^[a-z][a-z0-9+.-]*:/i.test(url) && !path.win32.isAbsolute(url)) {
+		const { protocol } = new urlUtils.URL(url);
+
+		if (protocol === "data:") {
+			fetchFromDataURL(url, (err, sourceContent) => {
+				if (err) return callback(err);
+				callback(null, { sourceURL: "", sourceContent });
+			});
+		} else if (protocol === "file:") {
+			const sourceURL = urlUtils.fileURLToPath(url);
+			fetchFromFilesystem(fs, sourceURL, (err, result) => {
+				if (err) return callback(err);
+				const { data: sourceContent } = result;
+				callback(null, { sourceURL, sourceContent });
+			});
+		} else {
+			callback(
+				new Error(`Failed to parse source map: '${url}' URL is not supported`)
+			);
+		}
+		return;
+	}
+
+	// 2. It's a scheme-relative
+	if (/^\/\//.test(url)) {
+		return callback(
+			new Error(`Failed to parse source map: '${url}' URL is not supported`)
+		);
+	}
+
+	// 3. Absolute path
+	if (isAbsolute(url)) {
+		const possibleRequests = [url];
+
+		if (url.startsWith("/") || url.startsWith("\\")) {
+			possibleRequests.push(
+				getAbsolutePath(fs, context, url.slice(1), sourceRoot)
+			);
+		}
+
+		fetchPathsFromFilesystem(fs, possibleRequests, "", (err, result) => {
+			if (err) return callback(err);
+			const { path: sourceURL, data: sourceContent } = result;
+			callback(null, { sourceURL, sourceContent });
+		});
+	} else {
+		// 4. Relative path
+		const sourceURL = getAbsolutePath(fs, context, url, sourceRoot);
+
+		fetchFromFilesystem(fs, sourceURL, (err, result) => {
+			if (err) return callback(err);
+			const { data: sourceContent } = result;
+			callback(null, { sourceURL, sourceContent });
+		});
+	}
+}
+
+/**
+ * @param {string} input input data
+ * @param {InputFileSystem} fs filesystem
+ * @param {NormalModule} module normal module
+ * @param {boolean} removeOnly should only remove source map comment
+ * @param {(err: Error|null, code?: string, sourceMap?: object) => void} callback callback
+ */
+module.exports = function extractSourceMap(
+	input,
+	fs,
+	module,
+	removeOnly,
+	callback
+) {
+	const { sourceMappingURL, replacementString } = getSourceMappingURL(input);
+
+	if (!sourceMappingURL) {
+		callback(null, input);
+		return;
+	} else if (removeOnly) {
+		callback(null, input.replace(replacementString, ""));
+		return;
+	}
+
+	fetchFromURL(
+		fs,
+		module.context || "",
+		sourceMappingURL,
+		"",
+		(err, result) => {
+			if (err) return callback(err);
+			const { sourceURL, sourceContent } = result;
+
+			if (sourceURL && module.buildInfo && module.buildInfo.fileDependencies) {
+				module.buildInfo.fileDependencies.add(sourceURL);
+			}
+
+			let map;
+
+			try {
+				map = JSON.parse(sourceContent.replace(/^\)\]\}'/, ""));
```

Comment:
should we:
- flatten source map if there is `map.sections`
- try to read source content for sources which doesn't have it in `map.sourceContent[i]`

Reply:
> * flatten source map if there is `map.sections`

That's really really uncommon. I would even stay nobody is using these types of SourceMaps.

> * try to read source content for sources which doesn't have it in `map.sourceContent[i]`

Yes, that would make sense. Also make sure to add these files as `this.addDependency`